tools/harness-simulation/posix/otbr_sim/otbr_docker.py

```diff
@@ -0,0 +1,170 @@
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+import logging
+import os
+import subprocess
+import time
+
+from config import (OT_PATH, OTBR_DOCKER_IMAGE)
+
+
+class OtbrDocker:
+
+    def __init__(self, nodeid: int, docker_name: str):
+        self.nodeid = nodeid
+        self.docker_name = docker_name
+
+        self.logger = logging.getLogger('otbr_docker.OtbrDocker')
+        self.logger.setLevel(logging.INFO)
+
+        self._socat_proc = None
+        self._ot_rcp_proc = None
+
+        self._rcp_device_pty = None
+        self._rcp_device = None
+
+        self._launch()
+
+    def __repr__(self) -> str:
+        return f'OTBR<{self.nodeid}>'
+
+    def _launch(self):
+        self.logger.info('Launching %r ...', self)
+        self._launch_socat()
+        self._launch_ot_rcp()
+        self._launch_docker()
+        self.logger.info('Launched %r successfully', self)
+
+    def close(self):
+        self.logger.info('Shutting down %r ...', self)
+        self._shutdown_docker()
+        self._shutdown_ot_rcp()
+        self._shutdown_socat()
+        self.logger.info('Shut down %r successfully', self)
+
+    def _launch_socat(self):
+        self._socat_proc = subprocess.Popen(['socat', '-d', '-d', 'pty,raw,echo=0', 'pty,raw,echo=0'],
+                                            stderr=subprocess.PIPE,
+                                            stdin=subprocess.DEVNULL,
+                                            stdout=subprocess.DEVNULL)
+
+        line = self._socat_proc.stderr.readline().decode('ascii').strip()
+        self._rcp_device_pty = line[line.index('PTY is /dev') + 7:]
```

Comment:
What is the magic number 7? Is it the length of `'PTY is `?

nit: I think regex will be more clear and understandable here.

Reply:
Use regex instead.